File: backend/district_geo.py
"""District centroids from bundled BharatMap GeoJSON (aligned with frontend map)."""
from __future__ import annotations

import json
from pathlib import Path
from typing import Dict, Optional, Tuple

ALIASES = {"Ahilyanagar": "Ahmednagar", "Mumbai": "Mumbai City"}
GEOJSON_PATH = Path(__file__).parent.parent / "frontend" / "public" / "data" / "maharashtra-districts.geojson"

_centroid_cache: Optional[Dict[str, Tuple[float, float]]] = None


def _ring_centroid(ring):
    n = max(len(ring) - 1, 1)
    lng = sum(p[0] for p in ring[:n]) / n
    lat = sum(p[1] for p in ring[:n]) / n
    return lat, lng


def _feature_centroid(feature: dict) -> Tuple[float, float]:
    geom = feature.get("geometry") or {}
    gtype = geom.get("type")
    if gtype == "Polygon":
        return _ring_centroid(geom["coordinates"][0])
    if gtype == "MultiPolygon":
        best, best_area = geom["coordinates"][0][0], 0.0
        for polygon in geom["coordinates"]:
            ring = polygon[0]
            area = abs(
                sum(
                    ring[i][0] * ring[(i + 1) % len(ring)][1]
                    - ring[(i + 1) % len(ring)][0] * ring[i][1]
                    for i in range(len(ring))
                )
            )
            if area > best_area:
                best_area, best = area, ring
        return _ring_centroid(best)
    return 19.3, 75.7


def load_district_centroids() -> Dict[str, Tuple[float, float]]:
    global _centroid_cache
    if _centroid_cache is not None:
        return _centroid_cache
    data = json.loads(GEOJSON_PATH.read_text(encoding="utf-8"))
    out: Dict[str, Tuple[float, float]] = {}
    for feat in data.get("features", []):
        props = feat.get("properties") or {}
        name = props.get("dtname") or props.get("New_Name")
        if not name:
            continue
        lat, lng = _feature_centroid(feat)
        out[name] = (round(lat, 4), round(lng, 4))
    for geo_name, seed_name in ALIASES.items():
        if geo_name in out:
            out[seed_name] = out[geo_name]
    _centroid_cache = out
    return out


def resolve_asset_coordinates(
    district_name: Optional[str],
    district_id: Optional[str] = None,
    asset_id: Optional[str] = None,
    latitude: Optional[float] = None,
    longitude: Optional[float] = None,
) -> Tuple[float, float]:
    """Return [lat, lng] snapped to district centroid when possible."""
    centroids = load_district_centroids()
    base = None
    if district_name and district_name in centroids:
        base = centroids[district_name]
    if base:
        h = sum(ord(c) for c in (asset_id or "")) % 10000
        j_lat = ((h % 100) - 50) * 0.00025
        j_lng = (((h // 100) % 100) - 50) * 0.00025
        return round(base[0] + j_lat, 6), round(base[1] + j_lng, 6)
    if latitude is not None and longitude is not None:
        return float(latitude), float(longitude)
    return 19.3, 75.7

File: frontend/src/lib/districtGeo.js
/** GeoJSON district name aliases (seed DB names → BharatMap boundaries). */
export const DISTRICT_NAME_ALIASES = {
  Ahmednagar: "Ahilyanagar",
  "Mumbai City": "Mumbai",
};

function ringCentroid(ring) {
  const n = Math.max(ring.length - 1, 1);
  let lat = 0;
  let lng = 0;
  for (let i = 0; i < n; i += 1) {
    lng += ring[i][0];
    lat += ring[i][1];
  }
  return [lat / n, lng / n];
}

export function featureCentroid(feature) {
  const { geometry } = feature;
  if (!geometry) return null;

  if (geometry.type === "Polygon") {
    return ringCentroid(geometry.coordinates[0]);
  }

  if (geometry.type === "MultiPolygon") {
    let best = geometry.coordinates[0][0];
    let bestArea = 0;
    geometry.coordinates.forEach((polygon) => {
      const ring = polygon[0];
      const area = Math.abs(
        ring.reduce((acc, point, idx) => {
          const next = ring[(idx + 1) % ring.length];
          return acc + point[0] * next[1] - next[0] * point[1];
        }, 0),
      );
      if (area > bestArea) {
        bestArea = area;
        best = ring;
      }
    });
    return ringCentroid(best);
  }

  return null;
}

/** Build { districtName: [lat, lng] } from districts GeoJSON features. */
export function buildDistrictCentroidMap(districtsGeo) {
  const map = {};
  (districtsGeo?.features || []).forEach((feature) => {
    const name = feature.properties?.dtname || feature.properties?.New_Name;
    const center = featureCentroid(feature);
    if (name && center) map[name] = center;
  });
  Object.entries(DISTRICT_NAME_ALIASES).forEach(([seedName, geoName]) => {
    if (map[geoName]) map[seedName] = map[geoName];
  });
  return map;
}

function hashId(id) {
  const s = String(id || "");
  let h = 0;
  for (let i = 0; i < s.length; i += 1) h = (h * 31 + s.charCodeAt(i)) % 100000;
  return h;
}

const MH_LAT = [15.55, 22.05];
const MH_LNG = [72.55, 80.95];

function inMaharashtra(lat, lng) {
  return lat >= MH_LAT[0] && lat <= MH_LAT[1] && lng >= MH_LNG[0] && lng <= MH_LNG[1];
}

/**
 * Resolve pin position for an asset: prefer BharatMap district centroid + small jitter.
 */
export function resolveAssetPosition(asset, centroidMap) {
  const districtKey = asset.district_name || asset.district_id;
  const base = centroidMap[asset.district_name] || centroidMap[districtKey];

  if (base) {
    const h = hashId(asset.id || asset.asset_id);
    const jLat = ((h % 100) - 50) * 0.00025;
    const jLng = (((h >> 7) % 100) - 50) * 0.00025;
    return [base[0] + jLat, base[1] + jLng];
  }

  const lat = Number(asset.latitude);
  const lng = Number(asset.longitude);
  if (Number.isFinite(lat) && Number.isFinite(lng) && inMaharashtra(lat, lng)) {
    return [lat, lng];
  }

  return null;
}
